source/extensions/common/aws/credential_providers/iam_roles_anywhere_x509_credentials_provider.cc

```
@@ -0,0 +1,258 @@
+#include "source/extensions/common/aws/credential_providers/iam_roles_anywhere_x509_credentials_provider.h"
+
+#include "source/common/common/base64.h"
+#include "source/common/tls/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Aws {
+
+using std::chrono::seconds;
+
+constexpr std::chrono::hours REFRESH_INTERVAL{1};
+constexpr std::chrono::seconds REFRESH_GRACE_PERIOD{5};
+constexpr uint64_t X509_CERTIFICATE_MAX_BYTES{2048};
+
+void CachedX509CredentialsProviderBase::refreshIfNeeded() {
+  if (needsRefresh()) {
+    refresh();
+  }
+}
+
+IAMRolesAnywhereX509CredentialsProvider::IAMRolesAnywhereX509CredentialsProvider(
+    Server::Configuration::ServerFactoryContext& context,
+    envoy::config::core::v3::DataSource certificate_data_source,
+    envoy::config::core::v3::DataSource private_key_data_source,
+    absl::optional<envoy::config::core::v3::DataSource> certificate_chain_data_source)
+    : context_(context), certificate_data_source_(certificate_data_source),
+      private_key_data_source_(private_key_data_source),
+      certificate_chain_data_source_(certificate_chain_data_source) {
+
+  auto provider_or_error_ = Config::DataSource::DataSourceProvider::create(
+      certificate_data_source_, context.mainThreadDispatcher(), context.threadLocal(),
+      context.api(), false, X509_CERTIFICATE_MAX_BYTES);
+  if (provider_or_error_.ok()) {
+    certificate_data_source_provider_ = std::move(provider_or_error_.value());
+  } else {
+    ENVOY_LOG(error, "Invalid certificate data source");
```

It's unclear to me what the desired behavior here is.
IIUC the current code "fails silently". Is this code being executed on the config update path? if so should the update be rejected instead?

---

I've now created an initialize method which will validate these. When I integrate in part 3 of the patch, initialize will be checked and this will prevent this credential provider from being loaded if any of the provided data sources are unable to be loaded.
For your information, there were other checks that would have prevented credentials from being returned from this provider in this case, but I think this is a better pattern and prevents any use of the provider upfront.